File: 1-week-C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool luhn_algoritm_check(long cart_number);
void define_card_type(long cart_number);

int main(void)
{
    // get a card number
    long cart_number = get_long("Enter card Number: ");

    // check luhn algoritm
    if (luhn_algoritm_check(cart_number))
    {
        // define card type
        define_card_type(cart_number);
    }
    else
    {
        printf("INVALID\n");
    }
}

bool luhn_algoritm_check(long cart_number)
{
    int sum = 0;
    int count = 0;

    while (cart_number > 0)
    {
        int last_number = cart_number % 10;

        if (count % 2 == 1)
        {
            last_number *= 2;
            if (last_number > 9)
            {
                last_number = (last_number % 10) + (last_number / 10);
            }
        }

        sum += last_number; // add to summ
        cart_number /= 10;  // dispose last number
        count++;
    }
    return (sum % 10 == 0);
}

void define_card_type(long cart_number)
{
    int length = 0;
    long temp_cart_number = cart_number;

    while (temp_cart_number > 0) // count digits
    {
        temp_cart_number /= 10;
        length++;
    }

    // define first 2 digits
    long first_two_digits = cart_number;
    while (first_two_digits >= 100)
    {
        first_two_digits /= 10;
    }

    // checking card
    if ((length == 15) && (first_two_digits == 34 || first_two_digits == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 16) && (first_two_digits >= 51 && first_two_digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && (first_two_digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: 6-week-python/sentimental-credit/credit.py
from cs50  import get_string

def luhn_check(card_number):
    total = 0
    reverse_card = card_number[::-1]

    for i in range(len(reverse_card)):
        num = int(reverse_card[i])

        if i % 2 == 1:
            num *= 2
            if num > 9:
                num -= 9
        total += num
    return total % 10 == 0

def get_card(card_number):
    cards = {
        "34": "AMEX",
        "37": "AMEX",
        "4": "VISA",
        "51": "MASTERCARD",
        "52": "MASTERCARD",
        "53": "MASTERCARD",
        "54": "MASTERCARD",
        "55": "MASTERCARD"

    }

    for card in cards:
        if card_number.startswith(card):

            if cards[card] == "VISA":
                if len(card_number) == 13 or len(card_number) == 16:
                    if luhn_check(card_number):
                        return cards[card]
                    else:
                        return "INVALID"
                else:
                    return "INVALID"


            return cards[card]

    return "INVALID"

def main():
    card_number = get_string("Number: ")
    print(get_card(card_number))

main()
